README.md
# JSmartable

> Jquery free plugin to provide dynamic collapsable responsive table

## Getting Started

### Quick start
JSmartable is a Jquery free plugin to provide dynamic collapsable responsive table. You can choose a specific breakpoint 
for each column you want to hide sor the specified breakpoint. It's very simple to use, and requires only jQuery v1.9.1+, 
Bootstrap CSS and Fontawesome v4 or v5.

Several quick start options are availables:

* Clone the repo: **git clone https://github.com/d4v1dm013/jsmartable**
* Install with npm: **npm install jsmartable**
* Install with yarn: **yarn add jsmartable**

### Demo
Demo : https://jsmartable.zestededev.com/

## Usage

You have to create a table with a **thead** and **tbody** section. For the plugin to work you must specify the for each **th** contained in the **thead** section the **data-breakpoint** you want to apply.

Below the list of possible breakpoint (you can mix different breakpoint in your **thead**) :

* **xs**: <= 480px
* **sm**: <= 576px
* **md**: <= 992px
* **lg**: <= 1200px
* **xlg**: <= 1400px

```html
<table class="table jsmartable">
    <thead>
        <tr>
            <th data-breakpoint="lg">ID</th>
            <th data-breakpoint="lg">1 Column</th>
            <th>2 Column</th>
            <th data-breakpoint="md">3 Column</th>
            <th data-breakpoint="md">4 Column</th>
            <th data-breakpoint="md">Lastname</th>
            <th data-breakpoint="xs">Firstname</th>
            <th data-breakpoint="xs">Title</th>
            <th data-breakpoint="lg">Description</th>
            <th>Action</th>
        </tr>
    </thead>
    <tbody>
        ...
    </tbody>
</table>
```

#### Via jsmartable class

Add the **jsmartable** class to your table to auto-initialize Jsmartable and configure each colunm to the specified breakpoint

```html
<table class="table jsmartable">
    <thead>
        <tr>
            ...
        </tr>
    </thead>
    <tbody>
        ...
    </tbody>
</table>
```

#### Via JavaScript

```javascript
// To style only selects with the my-custom-table class
 $('.my-custom-table').jsmartable();
```

or

```javascript
// To style all table
 $('table').jsmartable();
```

## Options

#### Global Options

You can customize some options if you want, below list of the possible options :


**allExpended** : Boolean `true` or `false`
*allow to open all lines by default*

**breakpoint** : `array` breakpoint: {  
	xs: `int` 480 by default,  
	sm: `int` 576 by default,  
	md: `int` 992 by default,  
	lg: `int` 1200 by default,  
	xlg: `int` 1400 by default,  
}
*allow to change the breakpoint options*

**iconPlus**: Code HTML, for exemple `<i class="fas fa-plus text-info"></i>`
*allow to change the plus icon*

**iconMinus** : Code HTML, for exemple `<i class="fas fa-minus text-info"></i>`
*allow to change the minus icon*

```javascript
// To style only selects with the my-custom-table class
$('.my-custom-table').jsmartable(
    breakpoint: {
        xs: 480,
        sm: 576,
        md: 992,
        lg: 1200,
        xlg: 1400,
    },
    iconPlus: '<i class="fas fa-plus text-info"></i>',
    iconMinus: '<i class="fas fa-minus text-info"></i>',
    allExpended: false
);
```
#### Rows Options

You can also open only some rows by default, for example the first row

```html
    <thead>
        <tr>
            ...
        </tr>
    </thead>
    <tbody>
        <tr data-opened="true">
            ...
        </tr>
            ...
    </tbody>
</table>
```

#### Columns Options

You can override the default title from the th column by a custom value. To do that, see below :

```html
    <thead>
        <tr>
            ...
        </tr>
    </thead>
    <tbody>
        <tr>
            <td data-title="Rewrite title 1"></td>
            <td></td>
            <td data-title="Rewrite title "></td>
            ...
        </tr>
            ...
    </tbody>
</table>
```

CHANGELOG.md
# v1.1.1 (2019-11-21)

### Bug Fixes
* Bug bad column

# v1.1.0 (2019-10-02)

### Bug Fixes
* Replace css icon class **fas** to **fa** for fontawesome 4 compatility

### New Features
* add missing peerDependencies to package.json

# v1.0.0 (2019-09-30)

### New Features
* Jquery free plugin to provide dynamic collapsable responsive table

jsmartable.js
$.fn.jsmartable = function(options){

    var settings = $.extend({
        breakpoint: {
            xs: 480,
            sm: 576,
            md: 992,
            lg: 1200,
            xlg: 1400,
        },
        iconPlus: '<i class="fa fa-plus text-info"></i>',
        iconMinus: '<i class="fa fa-minus text-danger"></i>',
        allExpended: false
    }, options);

    var tableSelector = this;

    $(window).resize(function() { 
        responsiveTable();
    }); 

    responsiveTable();

    function responsiveTable() {
        
        var screenSize = $(window).width();

        tableSelector.each(function(){

            var table = $(this);

            //all expanded
            var allExpanded = table.data('all-expanded');

            //If not defined in data we get settings from options
            if((typeof allExpanded == "undefined")) {
                allExpanded = settings.allExpended;
            }

            //Count nb cols has defined with data-breakpoint
            var th = table.find("thead tr:first th[data-breakpoint]");
            var totalColumns = th.length;

            //Return false if no column found
            if(totalColumns <= 0)
                return false

            //Get all rows contain in tbody
            var rows = table.find("> tbody > tr").not( ".jsmartable-row" );

            //loop all row to add a row responsive
            rows.each(function(index) {

                var row = $(this);
                var td = row.find("td");

                var formated = row.data('formated');

                //allow to expend automatically a row
                var expanded = row.data('expanded');

                var html = `<table class="jsmartable-subtable table table-bordered">
                                <tbody>`;

                var totalElement = 0;
                for(i=0;i<th.length;i++) {

                    //get current th information
                    var currentTh = th.eq(i);
                    var columnPosition = currentTh.prop("cellIndex")+1;
                    var breakpoint = currentTh.data('breakpoint');

                    //get current td information
                    var currentTd = row.find("td:nth-child("+(columnPosition)+")");
                    

                    //Allow to use a custom title instead of th
                    var title = currentTd.data('title');
                    if(typeof title == "undefined" || title == "") {
                        title = currentTh.html();
                    }

                    //Check if tf column exist at specified position
                    if(currentTd <= 0) {
                        console.log('TD Column ' + columnPosition + ' not found');
                        return false;
                    }

                    if(settings.breakpoint[breakpoint] != "undefined" && settings.breakpoint[breakpoint] >= screenSize) {
        
                        html += `<tr class="jsmartable-subrow">
                            <td class="jsmartable-subcol" style="width:30%; background:#eee;">${title}</td>
                            <td class="jsmartable-subcol">${currentTd.html()}</td>
                        </tr>`;

                        //Hide column selected
                        currentTh.hide();
                        currentTd.hide();
                        
                        totalElement++;
                    }
                    else {

                        //show column selected
                        currentTh.show();
                        currentTd.show();
                    }
                }

                html += `
                    </tbody>
                </table>`;

                if(typeof formated == "undefined" && totalElement > 0) { //Create jsmartable

                    row.data('formated', true);

                    row.after(`<tr id="jsmartable-row-${index}" class="jsmartable-row">
                        <td colspan='100%' class="jsmartable-col">
                            ${html}
                        </td>
                    </tr>`);

                    var buttonIcon = settings.iconMinus;
                    if((typeof expanded == "undefined" || expanded == false) && (typeof allExpanded == "undefined" || allExpanded == false)) {
                        $("#jsmartable-row-"+index).hide();
                        buttonIcon = settings.iconPlus;
                    }
                    
                    //Add button collapse
                    $("#jsmartable-row-"+index).prev().find('td:visible').first().prepend(`<a href='javascript:void(0);' class="mr-2 jsmartable-collapse" data-opened="true">${buttonIcon}</a>`);
                }
                else if(formated && totalElement > 0) { //update jsmartable

                    $('#jsmartable-row-'+index+' > td').html(html);
                }
                else if(totalElement <= 0) { //Remove jsmartable

                    row.removeData('formated');

                    $("#jsmartable-row-"+index).remove();

                    //remove button collapse
                    $(".jsmartable-collapse").remove();
                }
            });
        });
    }

    //Event collapse on click to open or close jsmartable
    $('body').on('click', '.jsmartable-collapse', function() {

        var $this = $(this);
        var jsmartableRow = $this.closest("td").closest("tr").next();

        //if Row is visible
        if (jsmartableRow.is(':visible')) {

            //hide row
            jsmartableRow.fadeOut("fast");
            
            //change the icon to plus
            $this.html(settings.iconPlus);
        } else {

            //show row
            jsmartableRow.fadeIn();

            //change the icon to minus
            $this.html(settings.iconMinus);
        }
    });
};

$(".jsmartable").jsmartable();